Settings safe integration

pheuberger opened the pull request:
This PR introduces a set of features to enable Safe multisig support to make changes to the profile name and avatar image in `/settings`. 

1. A selector modal that enables the user to select between their EOA and all the Safes the user is an owner of for the currently selected chain.
2. Display a pending update message when an update was initiated and is still to be signed by the other members of the Safe
3. Let's the user cancel the pending update request in case there are typos or the new data isn't in agreement with the other signers
4. TODO: allow user to refresh the pending request to see if the threshold of the Safe was met and commit the change to the database; also process the requests when logging in?

bitbeckers commented on lib/sign-api-message.ts:
```diff
@@ -0,0 +1,31 @@
+
+  // The type assertion to any is necessary because TS will otherwise complain,
+  // but the config is valid.
+  const signature = await signTypedData(wagmiConfig as any, {
```
The config is generated using `createConfig`. But the return type is probably dynamically infered?

Could we use the `useConfig` hook and provide the returned config as input to the method? 

https://wagmi.sh/react/api/hooks/useConfig

bitbeckers reviewed (commented):
Thanks! Mostly reviewed the business logic and changes in the code. As the build doesn't succeed, we'll need another pass when that's resolved.

pheuberger commented on components/settings/settings-form.tsx:
```diff
@@ -6,6 +6,7 @@ import Image from "next/image";
 import { useForm, useWatch } from "react-hook-form";
 import { zodResolver } from "@hookform/resolvers/zod";
 import { useAccount, useEnsAvatar, useEnsName } from "wagmi";
+import isURL from "validator/lib/isURL";
```
I was just using anything that works. Happy to do it using zod :+1:

pheuberger replied:
I was playing around with this some more and zod's url validation is a bit more permissive than I'd like. I can change the code so that it only tries to generate the preview when the user de-focuses the input field, which drastically improves the situation, but then, even entering a broken url like `http//example.org/image.png` (see the missing colon) crashes the app which is bad. `isURL` is a reliable option here that allows us to keep the rendering while the user types and the moment they produce a valid URL the image pops in. 

We could of course roll our own url validation with just a few lines of code. Let me know if you prefer that instead @bitbeckers.

bitbeckers replied:
Great, this is fine. Just wanted to make sure the step to not use zod had a reason

pheuberger commented on hooks/useSignAPIMessage.ts:
```diff
@@ -0,0 +1,48 @@
+        throw new Error("No chainId found");
+      }
+
+      const signature = await signTypedDataAsync({
```
If the user aborts the signature request, you'll see a toast with that error message. See this Loom https://www.loom.com/share/4faaa304f02c4c8bb25dcf956dae53de?sid=0651a549-164d-4eac-bc5b-d29eb9159f7c

bitbeckers commented on users/hooks.ts:
```diff
@@ -11,6 +11,9 @@ import { UserFragment } from "@/users/fragments/user.fragments";
 import request from "graphql-request";
 import revalidatePathServerAction from "@/app/actions";
 
+const STEP_1 = "step1";
```
Creates this issue for now, will refine a bit more during the board cleanup or feel free to add some specs here: https://github.com/hypercerts-org/hypercerts-app/issues/330

bitbeckers commented on hooks/useAccountDetails.ts:
```diff
@@ -0,0 +1,67 @@
+  ensName?: string | null,
+) {
+  if (ensName) return ensName;
+  if (accountType === "personal") return "Personal Account";
```
Leave this for now, Holke usually also QA's when we're live on testnet so if the language needs an update we can do that at tha time

pheuberger commented:
@bitbeckers I implemented the changes and fixed the bug I was facing. Everything builds locally and I re-deployed the Vercel preview. I hope the PR will recognize that.

bitbeckers commented:
> @bitbeckers I implemented the changes and fixed the bug I was facing. Everything builds locally and I re-deployed the Vercel preview. I hope the PR will recognize that.
> 
> Seems as though Vercel doesn't pick up my force-pushed changes, so when I redeploy it will redeploy the previously failing commit. I might have to create a new PR. Unless you have a better idea?

Just kicked the Vercel deployment from my end. If it indeed doesn't work, it'd need a new PR yeah

pheuberger commented:
> > @bitbeckers I implemented the changes and fixed the bug I was facing. Everything builds locally and I re-deployed the Vercel preview. I hope the PR will recognize that.
> > Seems as though Vercel doesn't pick up my force-pushed changes, so when I redeploy it will redeploy the previously failing commit. I might have to create a new PR. Unless you have a better idea?
> 
> Just kicked the Vercel deployment from my end. If it indeed doesn't work, it'd need a new PR yeah

It seems the build completed successfully. Feel free to merge any time :)
